gems/aws-sdk-s3/lib/aws-sdk-s3/file_downloader.rb

```diff
@@ -151,50 +124,63 @@ def multithreaded_get_by_parts(n_parts, total_size, etag)
 
       def download_in_threads(pending, total_size)
         threads = []
+        max_requests = pending.count
+        total_requests = 0
         progress = MultipartProgress.new(pending, total_size, @progress_callback) if @progress_callback
         @thread_count.times do
           thread = Thread.new do
             begin
-              while part = pending.shift
+              while (part = pending.shift)
                 if progress
                   part.params[:on_chunk_received] =
                     proc do |_chunk, bytes, total|
                       progress.call(part.part_number, bytes, total)
                     end
                 end
                 resp = @client.get_object(part.params)
+                validate_range(resp, part) if part.params[:range]
                 write(resp)
                 if @on_checksum_validated && resp.checksum_validated
                   @on_checksum_validated.call(resp.checksum_validated, resp)
                 end
+                total_requests += 1
               end
               nil
-            rescue => error
-              # keep other threads from downloading other parts
-              pending.clear!
-              raise error
+            rescue StandardError => e
+              pending.clear! # keep other threads from downloading other parts
+              raise e
             end
           end
           threads << thread
         end
         threads.map(&:value).compact
+        return if max_requests == total_requests
+
+        msg = "multipart download failed: expected #{max_requests} parts but received #{total_requests}"
+        raise MultipartDownloadError, msg
+      end
+
+      def validate_range(resp, part)
+        range = resp.content_range.split(' ').last.split('/').first
+        expected_range = part.params[:range].split('=').last
+        return if expected_range == range
+
+        raise MultipartDownloadError, 'multipart download failed: file integrity checked failed'
```

Comment:
I could be more descriptive - like "failed to validate content range of a part" or something.